src/model/element-base.ts

```diff
@@ -95,7 +131,180 @@ export abstract class ElementBase implements Feature {
    * single list. A mixin can be applied more than once, each time its
    * members override those before it in the prototype chain.
    */
-  mixins: Reference[] = [];
+  readonly mixins: ImmutableArray<Reference>;
+
+  constructor(init: ElementBaseInit, document: Document) {
+    const {
+      description,
+      summary,
+      demos,
+      events,
+      sourceRange,
+      properties,
+      attributes,
+      methods,
+      astNode,
+      warnings,
+      slots,
+      privacy,
+      jsdoc,
+      superClass,
+      mixins,
+    } = init;
+    this.description = description;
+    this.summary = summary;
+    this.demos = demos ? Array.from(demos) : [];
+    this.sourceRange = sourceRange;
+    this.astNode = astNode;
+    this.warnings = warnings ? Array.from(warnings) : [];
+    this.slots = slots ? Array.from(slots) : [];
+    this.privacy = privacy;
+    this.jsdoc = jsdoc;
+    this.superClass = superClass ? superClass.resolve(document) : undefined;
+    this.mixins = (mixins || []).map((m) => m.resolve(document));
+
+    const prototypeChain = this._getPrototypeChain(document, init);
+    for (const superClass of prototypeChain) {
+      this.inheritFrom(superClass);
+    }
+
+    this._overwriteInherited(
+        this.properties, properties || [], undefined, true);
+    this._overwriteInherited(
+        this.attributes, attributes || [], undefined, true);
+    this._overwriteInherited(this.methods, methods || [], undefined, true);
+    this._overwriteInherited(this.events, events || [], undefined, true);
+
+    for (const method of this.methods) {
+      // methods are only public by default if they're documented.
+      method.privacy = getOrInferPrivacy(method.name, method.jsdoc, true);
+    }
+  }
+
+  /**
+   * Returns the elementLikes that make up this class's prototype chain.
+   *
+   * Should return them in the order that they're constructed in JS
+   * engine (i.e. closest to HTMLElement first, closest to `this` last).
+   */
+  protected _getPrototypeChain(document: Document, _init: ElementBaseInit) {
+    const mixins = this.mixins.map(
+        (m) =>
+            this._resolveReferenceToSuperClass(m, document, 'element-mixin'));
+    const superClass = this._resolveReferenceToSuperClass(
+        this.superClass, document, 'element');
+
+    const prototypeChain: ElementBase[] = [];
+    if (superClass) {
+      prototypeChain.push(superClass);
+    }
+    for (const mixin of mixins) {
+      if (mixin) {
+        prototypeChain.push(mixin);
+      }
+    }
+
+    return prototypeChain;
+  }
+
+  protected _resolveReferenceToSuperClass(
```

Comment:
maybe? but that would be a bit tricky to do right because the reference would need to know the kind of thing that it's referencing. added a TODO to take a shot at this later.

I'm kinda thinking it should be something like `Reference<Kind>`?